main/uart_nic.c

```diff
@@ -133,12 +143,42 @@ static uint32_t IRAM_ATTR now_seconds() {
 
 static atomic_uint_least32_t last_inbound_seen = 0;
 static atomic_bool associated = false;
+static atomic_bool wifi_running = false;
 
 static bool beacon_quirk;
 static uint8_t probe_max_reties = 3;
 static atomic_bool probe_in_progress = false;
 static uint8_t probe_retry_count;
 
+typedef enum {
+    SCAN_TYPE_UNKNOWN,
+    SCAN_TYPE_PROBE,
+    SCAN_TYPE_LONG,
+    SCAN_TYPE_LOOPING_SHORT,
+    SCAN_TYPE_LOOPING_LONG,
+} ScanType;
+
+typedef struct {
+    uint8_t ssid[SSID_LEN];
+    bool needs_password;
+} ScanResult;
+
+typedef void (*wifi_scan_callback)(wifi_ap_record_t *, int);
+
+static struct {
+    ScanType scan_type;
+    wifi_scan_callback callback;
+    ScanResult stored_ssids[SCAN_MAX_STORED_SSIDS];
+    uint8_t stored_ssids_count;
+    volatile bool in_progress;
```

Comment:
The fact we need `volatile` scares me. Do you, by any chance, mean `atomic` instead?